### PR TITLE
Unify docs about .mjs, .cjs, .js and .json config files

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
Docs for https://github.com/babel/babel/pull/10903 and https://github.com/babel/babel/pull/10783, follow up to https://github.com/babel/website/pull/2106.

### CODE REVIEW

File: docs/config-files.md [hunk 1]
```diff
@@ -99,10 +88,46 @@ Unfortunately, this approach can be a bit repetitive, and depending on how Babel
...
+  However, the downside is that JS configs are less statically analyzable, and therefore have
+  negative effects on cacheability, linting, IDE autocomplete, etc.
+  Since `babel.config.json` and `.babelrc.json` are static JSON files, it allows other tools that
+  use Babel such as bundlers to cache the results of Babel safely, which can be a huge build
+  performance win.
+
+- `babel.config.cjs` and `.babelrc.cjs` allow you to define your configuration as CommonJS,
+  using `module.exports`.
+
+- `babel.config.mjs` and `.babelrc.mjs` use native ECMAScript modules. They are only supported by
```
Comment (existentialism) [inline docs/config-files.md]: They are supported by Node.js 13.2+ (or older versions via the `--experimental-modules` flag)

**Comment by existentialism** [comment]
Honestly not a fan of using `babel.config.*` everywhere... but I think enumerating all through all the various cases in one place is a needed improvement.

**Comment by nicolo-ribaudo** [comment]
What do you think about always using `babel.config.json` (or `babel.config.js`) and only mention in one place that there are different file extensions?

Btw, I copied `.*` from the ESLint docs (https://eslint.org/docs/user-guide/configuring)

File: docs/configuration.md [hunk 2]
```diff
@@ -70,9 +77,9 @@ Alternatively, you can choose to specify your [`.babelrc`](#babelrc) config from
 }
 ```
 
-### `.babelrc.js`
+### JavaScript configuratino files
```
Comment (jaydenseric) [inline docs/configuration.md]:
```suggestion
### JavaScript configuration files
```

**Comment by nicolo-ribaudo** [comment]
@existentialism I changed the docs to always use the `.json` extension instead `.*` (I chose `.json` because it's the simpler format). The different extensions are still explained in the `config-files.md` page.

File: docs/config-files.md [hunk 3]
```diff
@@ -8,16 +8,17 @@ id: config-files
...
+  * `.babelrc.json` files, with the different extensions
   * `package.json` files with a `"babel"` key
 
-
 ## Project-wide configuration
 
 New in Babel 7.x, Babel has a concept of a ["root"](options.md#root) directory, which defaults
 to the current working directory. For project-wide configuration, Babel will automatically search
-for a `"babel.config.js"`, `"babel.config.cjs"` or `"babel.config.json"` in this root directory. Alternatively, users can use an explicit
+for a `babel.config.json` file, or an equivalend one using the [supported extensions](#supported-file-extensions),
```
Comment (existentialism) [inline docs/config-files.md]:
```suggestion
for a `babel.config.json` file, or an equivalent one using the [supported extensions](#supported-file-extensions),
```

File: docs/config-files.md [hunk 4]
```diff
@@ -31,20 +32,9 @@ See the [monorepo](#monorepos) documentation for examples of how to use config f
...
-IDE autocomplete, etc. Since `babel.config.json` is a static JSON file, it allows other tools that use Babel
-such as bundlers to cache the results of Babel safely, which can be a huge build performance win.
-
-If you need a JavaScript configuration file, you have two options: `babel.config.js` or `babel.config.cjs`.
-They are almost equivalent, except when your `package.json` file contains the [`"type": "module"`](https://nodejs.org/api/esm.html#esm_code_package_json_code_code_type_code_field) option: since Babel configuration files must be scripts and not modules, in this case you must use `babel.config.cjs`.
-
 ## File-relative configuration
 
-Babel loads `.babelrc` (and `.babelrc.js` / `.babelrc.cjs` / `package.json#babel`) files by searching up the
+Babel loads `.babelrc.json` files, or an equivalend one using the [supported extensions](#supported-file-extensions), by searching up the
```
Comment (existentialism) [inline docs/config-files.md]:
```suggestion
Babel loads `.babelrc.json` files, or an equivalent one using the [supported extensions](#supported-file-extensions), by searching up the
```

File: docs/config-files.md [hunk 5]
```diff
@@ -99,10 +88,46 @@ Unfortunately, this approach can be a bit repetitive, and depending on how Babel
...
 
+## Supported file extensions
+
+Babel can be configured using any file extension natively supported by Node.js: you can use `.json`,
+`.js`, `.cjs` and `.mjs`, both for `babel.config.json` and `.babelrc.json` files.
+
+- `babel.config.json` and `.babelrc.json` are parsed as JSON5 and should contain an object matching
+  the [options](options.md) format that Babel accepts.
+  
+  It is recommended to choose this file type wherever possible: JS config files are
```
Comment (existentialism) [inline docs/config-files.md]:
```suggestion
  We recommend using this file type wherever possible: JS config files are
```

File: docs/config-files.md [hunk 6]
```diff
@@ -99,10 +88,46 @@ Unfortunately, this approach can be a bit repetitive, and depending on how Babel
...
+  handy if you have complex configuration that is conditional or otherwise computed at build time.
+  However, the downside is that JS configs are less statically analyzable, and therefore have
+  negative effects on cacheability, linting, IDE autocomplete, etc.
```
Comment (existentialism) [inline docs/config-files.md]: Sidenote, we might later want to illustrate more concrete examples of the effects (i.e. if you use .json with ____ then performance is much slower, etc).

File: docs/config-files.md [hunk 7]
```diff
@@ -99,10 +88,46 @@ Unfortunately, this approach can be a bit repetitive, and depending on how Babel
...
+- `babel.config.js` and `.babelrc.js` behave like the `.mjs` equivalents when your `package.json`
+  file contains the [`"type": "module"`](https://nodejs.org/api/esm.html#esm_code_package_json_code_code_type_code_field)
+  option, otherwise they are exactly the same as the `.cjs` files.
+
+JavaScript configuration files can either export an object, or a function that when called will
+return the generated configuration.
+Function-returning configs are given a few special powers because they can access an API exposed
+by Babel itself. See [Config Function API](#config-function-api) for more information.
+
+> For compatibility reasons, you could find `.babelrc` used as an alias for `.babelrc.json`.
```
Comment (existentialism) [inline docs/config-files.md]:
```suggestion
> For compatibility reasons, `.babelrc` is an alias for `.babelrc.json`.
```

**Pull request merged**